### PR TITLE
QUICKHOME for COREXY

### PR DESCRIPTION
**Pull request by darkjavi**
It comes from #2255 , updated fix, rebased and branched.

Trying to sum it up: On the sttepper.cpp there is an UPDATE_ENDSTOPS routine which is called in the homing process, that routine has a condition to check if the endstop is hit && the axis is mounted on is moving. That check used to be done by the step count of the stteper, which is not correct for COREXY because moving only one motor in corexy causes the head to move diagonally thus can hit either X endstop or Y endstop.

From the tree owners of COREXY participating the thread, only 2 where exposing the problem(@darkjavi and @rugbb) , @Salandora 's  machine is apparently quickhoming correctly though.

This pr exposes dx and dy(actual axis movement) on the block struct, in the same fashion of the steps array so in the UPDATE_ENDSTOP routine can replace :  

``` C
current_block->steps[_AXIS(AXIS)]
```

with

``` C
current_block->axis_movement[_AXIS(AXIS)] 
```

in case COREXY is enabled.

I tried this fix(not this particular pr yet,which is only compiled) last week and so far is working.It will be great if @ruggb, @Salandora, or any other COREXY owner could try it and give feedback about it.

### CODE REVIEW

**Comment by ruggb** [comment]
I have seen a number of changes. Just to be clear can u give me all the chgs I need to make to try this
OR is this already in a file(s)??

**Comment by darkjavi** [comment]
I think you can download the full zip here : 
https://github.com/darkjavi/Marlin/archive/quickhome-fix.zip

**Comment by ruggb** [comment]
well:
The polarity of the endstops is now reversed. I would take issue with you that this is the most common config. I am using a std PCB endstop that apparently a lot of ppl use that Mendal designed and the line is connected to ground via the NO contact. NORMAL config for input logic is negative - active = ground. I have been a h/w engineer for 60 yrs.

If u really want to make a chg like this my suggestion would be to have a single define up front that sets them all based on NO or NC to ground. A lot of less technical ppl who upgrade will have problems with this.

Otherwise the QH function appears to work.

Now, why is the default for the steppers ON. They also do not turn off after 60 sec idle.
So how do I set this back to the way it was????????

I really can't see a good reason for making the above changes but I am a newbee at this 3D printing.
It will just add confusion to upgraders, which I am sure they will love u for.

**Comment by tonnico** [comment]
The polarity isn't changed. This patch don't do anything with the endstop logics.

You need to use your config and not the standard one!

**Comment by ruggb** [comment]
it was chgd in 1.0.3 from 1.0.2 - I guess I implied the QH fix changed it.

I still need to know why the stepper timeout was changed. They do not come on at power on as I thought, but they do stay powered after a move for MUCH longer than 60 sec. as 1.0.2 does.
20 min and counting.............
I think it does eventually time out but apparently this was changed n 1.0.3 dev

I just dnld today's version of 1.0.3 dev and QH works in that also.

There is no gotcha for babystepping in corexy, so I am assuming it works for CoreXY now. True?

File: Marlin/stepper.cpp [hunk 1]
```diff
@@ -490,14 +490,24 @@ ISR(TIMER1_COMPA_vect) {
...
+          SET_ENDSTOP_BIT(AXIS, MINMAX); \
+          if (TEST_ENDSTOP(_ENDSTOP(AXIS, MINMAX))  && (current_block->steps[_AXIS(AXIS)] > 0)) { \
+            endstops_trigsteps[_AXIS(AXIS)] = count_position[_AXIS(AXIS)]; \
+            _ENDSTOP_HIT(AXIS); \
+            step_events_completed = current_block->step_event_count; \
+          }	  
+      #else // CoreXY checks for axis movement, not steps.
+	    #define UPDATE_ENDSTOP(AXIS,MINMAX) \
+	  	  SET_ENDSTOP_BIT(AXIS, MINMAX); \
+		  if (TEST_ENDSTOP(_ENDSTOP(AXIS, MINMAX))  && (current_block->axis_movement[_AXIS(AXIS)] != 0)) { \
```
Comment (tonnico) [inline Marlin/stepper.cpp:504]: compare float with zero is not the best idea.

**Comment by thinkyhead** [comment]
Just curious… Is the planner block structure moving closer to the planner block structure of other firmwares, such as Repetier, and is this how they also handle CoreXY? And, on another note, would this also be helpful for DELTA, SCARA, and other machines with kinematic translation?

File: Marlin/stepper.cpp (hunk 1 above)
Comment (darkjavi) [inline Marlin/stepper.cpp:504]: Oh, I was unaware of that. I see on stackoverflow that comparing float to 0 is unrealiable depending on the compiler, as for some   -0.0f != +0.0f, is that what you are pointing out?

Maybe I can try something like:

``` C
abs(current_block->axis_movement[_AXIS(AXIS)]) > 0
```

**Comment by darkjavi** [comment]
Well, it depends on the mechanical construction of the kinematic, SCARA and DELTA constructions has endstops triggered by the arms(A,B,C), so for them makes sense to fetch from the raw stepper.
Regular cartesian could either do steps or head direction, and COREXY has to do it with head direction.

@ruggb I'll have to test latest dev to see, does your machine have same size for x/y? (Xmax == Ymax)

**Comment by ruggb** [comment]
my bed size is 335x305 == the build size is 305x305. the extra 30 is for dual extruders
f/w is set at 335 x 305

File: Marlin/stepper.cpp (hunk 1 above)
Comment (thinkyhead) [inline Marlin/stepper.cpp:504]: We really should have some helpers for the purpose:

```
#define EPSILON ((float)0.0000001)
inline bool nearly_zero(float x) { return fabs(x) < EPSILON; }
inline bool nearly_equal(float x, float y) { return nearly_zero(x - y); }
```

Giving us…

```
if (TEST_ENDSTOP(_ENDSTOP(AXIS, MINMAX)) && !nearly_zero(current_block->axis_movement[_AXIS(AXIS)])) { ...
```

**Comment by darkjavi** [comment]
Hi there, I'm back from holidays! well, actually I have some 'not doing tourist stuff days' on the holidays while I wait in Berlin for the CCC camp next week.

I have my printer deployed in a friend's house and I can work on the issue these days and give it some testing.
Should I rework the thing to include the float helpers?
I see we are moving to a new dev repo, so I guess I'll have to rebase it from the new repo.

BTW, anyone going to CCC?

**Comment by Wackerbarth** [comment]
Yes, not only is this commit now quite a bit behind the "Development" to which you made the PR, but "master" over on MarlinDev has even more additions.  The good news is that both repositories have the same root.  Just add the MarlinDev as an additional remote on your local machine and fetch  those branches.  Rebasing should be no different than it was here.  You will need to submit a new PR over there and close this one.

**Comment by thinkyhead** [comment]
@Wackerbarth Do you have a good list of git command examples that can help explain how to (a) work with remotes and local branches, keeping things up to date? And (b) do you have any snippets for making PRs on top of other people's PR branches? I feel like I'm doing too much work sometimes on (a) and I can never remember how to deal with (b). Perhaps you can just pull up your BASH history and paste the typical git commands you've been using….

Meanwhile, I'm doing this: http://pcottle.github.io/learnGitBranching/

**Pull request closed without merging**